### PR TITLE
Validation pass

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #

## Proposed Changes
- added validation and test logic into the module
- implemented proper output saving, checkpointing, early stopping
- extended existing tests for all tasks and added cases for 3D radiology

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

### CODE REVIEW

File: GANDLF/models/lightning_module.py [hunk 1]
```diff
@@ -38,15 +44,20 @@
...
         super().__init__()
         self.output_dir = output_dir
         self.params = deepcopy(params)
         self.current_best_loss = sys.float_info.max
         self.wait_count_before_early_stopping = 0
-        self._problem_type_is_regression_or_classification = (
-            self._check_if_regression_or_classification()
+        self._problem_type_is_regression = params["problem_type"] == "regression"
+        self._problem_type_is_classification = (
+            params["problem_type"] == "classification"
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py:68]: These should all be mutually exclusive. Do you think it would be better for us to put this in a constant and if-else loop?

File: GANDLF/models/lightning_module.py [hunk 2]
```diff
@@ -136,12 +179,14 @@ def on_train_start(self):
...
+        self._initialize_train_logger()
         self._initialize_training_epoch_containers()
 
+        # TODO check out if the disbled by default medcam is indeed what we
+        # meant - it was taken from original code
         if "medcam" in self.params:
-            self._enable_medcam()
-
+            self._inject_medcam_module()
+            self.params["medcam_enabled"] = False
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py]: Why disabled?

File: GANDLF/models/lightning_module.py [hunk 3]
```diff
@@ -532,22 +580,438 @@ def _print_total_training_time(self):
...
+        )
+        self._ensure_path_exists(self._current_validation_epoch_save_dir)
 
-    # TODO placeholder for now
     def validation_step(self, subject, batch_idx):
-        loss = torch.randn([1])
+        if self.params["verbose"]:
+            self._print_currently_processed_subject(subject)
+
+        # TODO this is going to block any parallelism, as the spacing is going to unpredicatably change across GPUs
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py]: The idea behind this to allow the possibility for a model to take the subject's spacing into account during computation. IMO, as long as the **entire** same subject is present in the same GPU, it will be fine.

File: GANDLF/models/lightning_module.py [hunk 4]
```diff
@@ -532,22 +580,438 @@ def _print_total_training_time(self):
...
+                self.rows_to_write = []
+
+    @rank_zero_only
+    def _ensure_path_exists(self, path):
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+    # TODO called at the validation step, NOT at the end of the epoch - we want to avoid
+    # saving all predictions for all subjects for the end of the epoch
+    def _save_predictions_for_segmentation_subject(
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py:1299]: IMO this logic can go a separate submodule, right?

File: GANDLF/models/lightning_module.py [hunk 5]
```diff
@@ -532,22 +580,438 @@ def _print_total_training_time(self):
...
+            torchio.Subject(subject_dict),
+            self.params["patch_size"],
+            patch_overlap=self.params["inference_mechanism"]["patch_overlap"],
+        )
+        patch_loader = torch.utils.data.DataLoader(grid_sampler, batch_size=1)
+
+        return grid_sampler, patch_loader
+
+    # TODO will also suffer from the same issue as the training step with
+    # synchronization across GPUs
```
Comment (sarthakpati) [inline GANDLF/models/lightning_module.py]: Why would this suffer from sync issues?

**Review by sarthakpati — COMMENTED** [review]
Minor comments/questions. But one major: all the internal functions need docstring comments to make their functionality clear.

File: GANDLF/models/lightning_module.py (hunk 1 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py:68]: Yeah, they indeed are exclusive. Not really sure about the solution you mean, but probably result will be the same - one of them set to true, rest automatically to False. Also I think that here we should be assured that the config parsing handled that properly (it should) and mutual exclusiveness already happened.

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Well theoretically yes, but note that we have the "spacing" saved in params, which in this case is a class variable, and is used by all parallel processes. Therefore, each of them gets unique subject, and tries to write spacing into the self.params. And the value that is written will come from the process that finished the write last. It can be solved quite easlity tho I would probably need to modify the parst of original gandlf code that uses the "spacing" param

File: GANDLF/models/lightning_module.py (hunk 5 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: In DDP setting, each process would execute the validation/training/test epoch end logic separately - I am not sure how this would look like, I am always cautious of some strange race or parallel access conditions occuring. Preferably, after the port works, I will take this entire logic and execute it only on rank 0 process, as some of the used methods already do. Maybe we would be safe without doing so, but this seems like something that should be managed quite easily.

File: GANDLF/models/lightning_module.py (hunk 4 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py:1299]: Yes, I want to extract the entire saving logic and some others at second stage of port (when we have baseline working) to a separate classes and compose them into the main lightning module. Tho for now this solution was simpler for me to capture everything in one place.

File: GANDLF/models/lightning_module.py (hunk 5 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py]: Ah - I understand. Perhaps this information something we should in the comments.

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py]: Understood. 

If you look at [these lines](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/data/ImagesFromDataFrame.py#L141-L145) from the current GaNDLF code, we can see that the `"spacing"` key is being defined for each subject once. What happens in the [training loop](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/compute/training_loop.py#L131-L135) is a use case where other libraries (such as OpenFL) invoke the training loop directly with their dataloaders initialization (not an ideal scenario). I believe having this line would suffice, and _should_ not result in a race condition:

```python
subject["spacing"] = subject.get("spacing", None)
```

Any downstream use of `"spacing"` can use this entry directly and not `params["subject_spacing"]`.

Thoughts?

File: GANDLF/models/lightning_module.py (hunk 5 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Actually now I am working on resolving the docstrings, and I think I have simple solution for that. I am adding it, it will not change the single-device logic by any means. I removed those comments and left it one to further ingvestigate multi-gpu setup.

File: GANDLF/models/lightning_module.py (hunk 1 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py:68]: Ah, I see. I thought you are incorporating the logic from [the utils](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/utils/parameter_processing.py#L53-L125) (side note: I believe these functions can be combined into 1 for simplicity). 

And here is what I meant by the constant: we define an `enum` in the (currently non-existent) `GANDLF.constants` submodule (perhaps including the problem type determination logic from above as well):

```python
from enum import Enum

class ProblemType(Enum):
    SEGMENTATION = 0
    CLASSIFICATION = 1
    REGRESSION = 2
```

And then use it in place of `params["problem_type"]`. But I guess that be too much of an overkill, and perhaps left to the [pydantic check](https://github.com/mlcommons/GaNDLF/pull/976)?

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Totally agree, if we keep it that way each process does not have to care about the race.

Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Actually, in our case, the changes would have to be done in one function only (_calculator_generic_all_surface_distances). However, this one does not take any subject as an argument, should I temporairly modify it?

File: GANDLF/models/lightning_module.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: I did it accordingly to the original code - take look into the lines 409 to 421 in training_loop.py - it just injects the medcam module and sets medcam enabled to false. I wrote about this in a comment in  the lightning_module.py in method on_validation_epoch_start.

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py]: Yup, that makes sense. Thanks.

File: GANDLF/models/lightning_module.py (hunk 1 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py:68]: I see now, still tho, in that case I would opt for an approach that we merge it into single method but leave the class varaibles as is - the name would be much more expressive  (when variable has name like "_problem_type_is_regression" etc you instantly know what it means, and you can access that from whenever in the class code. Also not sure if I get how the Enum type would be helpful here, can you show some example with initialization of this class and later checking what type of problem we have now?

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Ok, after codebase search I see that the function calling "subject_spacing" is used only in the generate_metrics.py file. (Function name is  _calculator_generic_all_surface_distances), therefore I doubt that we need the subject spacing for any metric during the training or inference.Can you confirm? If so then I will simply delete the spacing settings and that is going to be it

File: GANDLF/models/lightning_module.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: @sarthakpati toughts on this?

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: @sarthakpati also here, am I correct?

Comment (sarthakpati) [reply GANDLF/models/lightning_module.py]: Sorry, I somehow missed the notification.

For now, that is correct. However, I think it would be better to have it since we might need this down the road. Especially if someone wants to use a distance-based loss function. 

Thoughts?

**Comment by sarthakpati** [comment]
Could you also fix the typos reported by the spelling checker?

File: GANDLF/models/lightning_module.py (hunk 2 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py]: Oh okay, then.

File: GANDLF/models/lightning_module.py (hunk 1 above)
Comment (sarthakpati) [reply GANDLF/models/lightning_module.py:68]: Precisely. 

But let's table this for now (and link this comment in the code somewhere for us to refer). I believe this can be combined with the pydantic check.

Comment (szmazurek) [reply GANDLF/models/lightning_module.py:68]: Cool, thanks!

File: GANDLF/models/lightning_module.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/models/lightning_module.py]: Ok, but maybe we should keep it there with the note that it should be a variable within the context of this method, not inserted in global params dict, because as I said I am worried that when we run in mulit-gpu setting we can get some strange race conditions (one process assigns the spacing for subject it analyzes, and then other process reads it for processing of separate subject). What you think?

Comment (sarthakpati) [reply GANDLF/models/lightning_module.py]: Yup, that sounds good.
